FIX HistGradientBoosting raising ValueError with monotonic_cst and categorical_feature

yuanx749 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #28898

#### What does this implement/fix? Explain your changes.
`HistGradientBoosting` uses `ColumnTransformer` to preprocess the input X, which places the categorical features at the beginning of X. This PR adjusts the feature order in `monotonic_cst` correspondingly, to avoid the error when calling `TreeGrower`.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

yuanx749 committed:
Add test

yuanx749 committed:
FIX HistGradientBoosting with monotonic_cst and categorical_feature

yuanx749 committed:
Add whatsnew

yuanx749 committed:
Merge remote-tracking branch 'upstream/main' into hist-mono

yuanx749 commented:
Thank you for reviewing @ogrisel. I added the entry to whats_new.

ogrisel commented:
Actually, I think it would be worth making the test stronger by expanding `test_predictions` to insert some extra columns `f_cat_a = rng.randint(low=0, high=9, size=n_samples)` (and similarly `f_b` and `f_c`, before, between and after `f_0` and `f_1` and no monotonicity constraint on those columns (and no impact on `y` either).

This way we could better check the correctness of the monotonicity constraint remapping.

ogrisel commented:
/cc @thomasjpfan.

yuanx749 committed:
Expand test_predictions

yuanx749 commented:
> Actually, I think it would be worth making the test stronger by expanding `test_predictions` to insert some extra columns `f_cat_a = rng.randint(low=0, high=9, size=n_samples)` (and similarly `f_b` and `f_c`, before, between and after `f_0` and `f_1` and no monotonicity constraint on those columns (and no impact on `y` either).
> 
> This way we could better check the correctness of the monotonicity constraint remapping.

Good point. I updated `test_predictions`, and all tests pass locally.

Then I think the additional non-regression test I added before can be removed.

thomasjpfan reviewed (approved):
Thanks for the PR @yuanx749 ! The fix and additional tests looks good to me.

ogrisel commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_monotonic_contraints.py:
```diff
@@ -433,3 +444,13 @@ def test_bounded_value_min_gain_to_split():
         upper_bound=children_upper_bound,
     )
     assert split_info.gain > min_gain_to_split
+
+
+def test_with_monotonic_cst_and_categorical_features():
+    # Non-regression test for #28898
+    X = np.arange(9).reshape(3, 3)
+    y = np.array([0, 1, 0])
+    hist = HistGradientBoostingClassifier(
+        monotonic_cst=[1, 0, -1], categorical_features=[False, True, False]
+    )
+    hist.fit(X, y)
```
```suggestion
```

ogrisel reviewed (approved):
Let's delete the redundant test but otherwise LGTM! Thanks for the fix.

ogrisel committed:
Remove redundant test

ogrisel committed:
Merge branch 'main' into hist-mono
